crypto/fipsmodule/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,623 @@
+/*
+ * Copyright 2020-2021 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2020-2021, Intel Corporation. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ *
+ *
+ * Originally written by Sergey Kirillov and Andrey Matyukov.
+ * Special thanks to Ilya Albrekht for his valuable hints.
+ * Intel Corporation
+ *
+ */
+#ifndef RSAZ_512_ENABLED
+/*
+ * This appears in openssl as just:
+ *
+ *   NON_EMPTY_TRANSLATION_UNIT
+ *
+ * This macro is defined in include/openssl/macros.h like so:
+ *
+ *   // Sometimes OPENSSL_NO_xxx ends up with an empty file and some
+ *   // compilers don't like that.  This will hopefully silence them.
+ *   # define NON_EMPTY_TRANSLATION_UNIT static void *dummy = &dummy;
+ */
+static void *__dummy = &__dummy;
+#else
+
+#include <openssl/opensslconf.h>
+#include <openssl/crypto.h>
+#include <assert.h>
+#include <string.h>
+#include "../../internal.h"
+#include "rsaz_exp.h"
+
+/* Internal radix */
+# define DIGIT_SIZE (52)
+/* 52-bit mask */
+# define DIGIT_MASK ((uint64_t)0xFFFFFFFFFFFFF)
+
+# define BITS2WORD8_SIZE(x)  (((x) + 7) >> 3)
+# define BITS2WORD64_SIZE(x) (((x) + 63) >> 6)
+
+/* Number of registers required to hold |digits_num| amount of qword digits */
+# define NUMBER_OF_REGISTERS(digits_num, register_size)            \
+    (((digits_num) * 64 + (register_size) - 1) / (register_size))
+
+OPENSSL_INLINE uint64_t get_digit(const uint8_t *in, int in_len);
+OPENSSL_INLINE void put_digit(uint8_t *out, int out_len, uint64_t digit);
+static void to_words52(BN_ULONG *out, int out_len, const BN_ULONG *in,
+                       int in_bitsize);
+static void from_words52(BN_ULONG *bn_out, int out_bitsize, const BN_ULONG *in);
```

Comment:
these functions work only when `BN_ULONG` is `uint64_t` so why not make that explicit and get rid of `BN_ULONG`?

Reply:
I think the comment preceding the definition, at least for `to_words52`, may be misleading. IIUC, because of the way elements of `BIGNUM`'s `d` member are packed, I think it will work whether `BN_BITS2` is 32 or 64. Unfortunately though, for `from_words52`, you're right. It assumes `BN_ULONG` is `uint64_t`. I've updated its definition and declaration.